beginner-workshop_rolling-update/src/exercises/Solutions.hs

```diff
@@ -0,0 +1,331 @@
+  | x == 1 = 1
+  | otherwise = x * factorial (x-1)
+
+
+
+
+
+-- #############################################################################
+-- Mergesort!
+-- #############################################################################
+
+mergesort :: Ord a => [a] -> [a]
+mergesort xs
+  | length xs == 0 = []
+  | length xs == 1 = xs
+  | otherwise = let (ys, zs) = splitIntoHalves xs
+    in merge (mergesort ys) (mergesort zs)
+  where
+    -- Merge two ordered lists into one
+    merge :: Ord a => [a] -> [a] -> [a]
+    merge (x:xs) (y:ys) = if x < y then x:(merge xs (y:ys)) else y:(merge (x:xs) ys)
+    merge [] ys = ys
+    merge xs [] = xs
+
+    -- Split a list into equally sized halves (± 1 element)
+    splitIntoHalves :: [a] -> ([a], [a])
+    splitIntoHalves (xs) = splitAt (div (length xs) 2) xs
+
+
+-- Get the first element matching the predicate
+find' :: (a -> Bool) -> [a] -> Maybe a
+find' predicate (y:ys) = if predicate y then Just y else find' predicate ys
+find' predicate [] = Nothing
+
+
+-- Pair entries up, stop when one list is empty
+-- e.g. zip' [1..] "abc" ==> [(1,'a'), (2,'b'), (3,'c')]
+zip' :: [a] -> [b] -> [(a,b)]
+zip' (x:xs) (y:ys) = [(x, y)] ++ (zip' xs ys)
+zip' [] _ = []
+zip' _ [] = []
+
+-- combine list elements using a function
+--
+-- >>> zipWith (+) [1,2,3] [10,20,30]
+-- [11,22,33]
+--
+-- NB: zip = zipWith (,)
+--     zipWith' f = map f . zip
+zipWith' :: (a -> b -> c) -> [a] -> [b] -> [c]
+zipWith' f (x:xs) (y:ys) = [(f x y)] ++ zipWith f xs ys
+zipWith' _ [] _ = []
+zipWith' _ _ [] = []
+
+
+-- Apply each function to the value
+--
+-- >>> applyAllTo 3 [even, odd, (> 3)]
+-- [False, True, False]
+applyAllTo :: a -> [a -> b] -> [b]
+applyAllTo = _todo
+
+
+
+
+-- #############################################################################
+-- IO
+-- #############################################################################
+
+-- »Run all the actions contained in a list.«
+--
+-- >>> runAll [putStrLn "hello", putStrLn "world"]
+-- hello
+-- world
+runAll :: [IO a] -> IO ()
+runAll [] = return ()
+runAll (x:xs) = do 
+  _ <- x
+  runAll xs
+  return ()
+
+
+
+-- Given
+-- Use `fmap` for IO to implement this function!
+-- Useful helpers:
+--   - readFile :: String -> IO String
+--   - length :: String -> Int
+getFileLength :: FilePath -> IO Int
+getFileLength name = ((fmap length) . readFile) name
+
+
+
+
+
+-- #############################################################################
+-- Laziness
+-- #############################################################################
+
+
+-- Keep only every other element of a possibly infinite list.
+everyOther :: [a] -> [a]
+everyOther (x:y:xs) = [x] ++ everyOther xs
+everyOther _ = []
+
+-- Split a list into halves of equal length (±1).
+-- Easy using `length`, tricky so it works even on infinite lists.
+splitHalf :: [a] -> ([a], [a])
+splitHalf (x:y:xs) = let (xs', ys') = splitHalf xs
+  in (([x]++xs'), ([y]++ys'))
+splitHalf _ = ([], [])
```

Comment:
Correct solution, the question is a bit fuzzy. 

What I meant was a function splitting the list "in the middle", splitHalf [1,2,3,4]=([1,2],[3,4])